### ISSUE TITLE
Request: configurable timeout in documentation building code.

### ISSUE THREAD

**Issue by azarens-git** [coq/coq#13848]
## Background
 
As it was discovered in the discussion thread [here](https://github.com/coq/coq/issues/7829#issuecomment-776528545), the build of documentation in Coq could fail if the coqc binary is working a bit too slow.  (cf. [these observations](https://github.com/coq/coq/issues/7829#issuecomment-776743836) from the same discussion thread).

There exist legitimate cases for such slow performance of coqc binary, in particular if it was compiled with an OCaml compiler whence a `spacetime` option is enabled.  In fact, the test results referenced thereby are for one such binary (also, cf. [these additional observations](https://github.com/coq/coq/issues/11178#issuecomment-777428518) ).

#### Coq Version

8.13.0
NetBSd 9.1_STABLE

## Request

I would like to request that the timeout for the documentation builds be made configurable, and that it also be made configurable for dune-based builds.

#### Additional notes

I could make myself available to implement these features, particularly concerning the part of dune-based builds.  I would appreciate an opportunity to learn more about the dune build system.

**Comment by Zimmi48** [comment]
Nice! What do you think would be a good way to change the default timeout?

Right now the timeout is set at the following line:
https://github.com/coq/coq/blob/c0e0e637c61e075f43b73d1ddd8eaa9d79b27561/doc/tools/coqrst/repl/coqtop.py#L72

But it could be made to depend on a variable defined in https://github.com/coq/coq/blob/master/doc/sphinx/conf.py or on an environment variable instead.

The former allows depending on a configure-time-generated `coq_config` module, if that's the way you want to use to increase the timeout (but this is less flexible since `configure` is usually run very early in the process).

The latter seems both simpler to implement and more flexible. If you go this way, you'll have to explicitly list this variable here so that it is picked by the Dune build:
https://github.com/coq/coq/blob/c0e0e637c61e075f43b73d1ddd8eaa9d79b27561/doc/dune#L20

**Comment by azarens-git** [comment]
I would prefer an option that is most harmonious with **dune**, and in particular, something that could be set as `dune` runs and executes its configuration routines under _config/_ directory.

1. I think, there's a module that `dune` produces, called `config`, under the directory _config/_.  Would it be easy to query timeout from there?
2. If not, then perhaps have at how old `configure.ml` produces a `coq_config.py` file with a snippet within it.  `dune` could produce another small python file with configuration.  Then, the source file should be able to simply include it.
3. I vote strongly against environment variables, since the environment in general should not be polluted.

**Comment by Zimmi48** [comment]
`config/` is (*only*) for the output of `configure.ml`, which is called by *both* the legacy build system and Dune. Right now, Dune only lets you tweak this by calling `configure` manually (which you can do), so I wouldn't call this "harmonious with dune". Environment variable can be temporary (only set on the command launching the build). It's fine if that's not what you want, but I wouldn't buy your environment pollution argument.

**Comment by azarens-git** [comment]
Could we use a `py.in` file with a snippet of code, which dune modifies by substitution of `@SHPINXWARNOPT@`?  It'll be just one more `run` instruction.

**Comment by azarens-git** [comment]
Of course, substitution could be performed on original source, which would at the time of substitution have a name `coqtop.py.in`.  It's a very classical method, and could be performed easily from dune through e.g. `sed`, if nothing else (alternatively, a binary to perform substitution could be compiled).

**Comment by cpitclaudel** [comment]
Why not just make the timeout a minute or so?  How often has that timeout caught bugs where Coq froze? (if I was the was who put it in, it was probably to debug my code as I started working on parsing Coq's output)

**Comment by azarens-git** [comment]
A minute-long timeout would be more than adequate, for e.g. a case of
compiling the code with special compiler flags that do make it
perform slowly only certain operations.  In particular, the `spacetime`
flags do cause a slowdown even on a machine with 128Gb of RAM.  In
practice, the slowdown would not exceed about 20 seconds on a
well-configured machine.  On a machine with less efficient setup involving
swap, a slowdown around 80 seconds had been observed.

**Comment by Zimmi48** [comment]
Just consider how long the total documentation build time will be on a machine where each (!) coqtop invocation takes dozens of seconds. I think a user might want to discover about this slowness issue with a timeout error rather than waiting for a very long time (and complaining) that the documentation takes so long to build.

**Comment by azarens-git** [comment]
If you're concerned about interactive builds, then let me suggest for you
to consider a warning to be issued.

As for non-interactive package builds, in those cases, the packages are
built in bulk, and the extra 30 minutes in a parallelized build setup would
not matter.  When it starts to matter, then a sysadmin is going to take
node and investigate.

At any rate, the longer builds result from special compiler setups (e.g.
having spacetime), and it is reasonable for someone who does these builds
to in fact know what he is doing.
